java/ql/lib/semmle/code/java/security/CleartextStorageQuery.qll

```diff
@@ -0,0 +1,108 @@
+/** Provides classes and predicates to reason about cleartext storage vulnerabilities. */
+
+import java
+private import semmle.code.java.dataflow.DataFlow4
+private import semmle.code.java.dataflow.TaintTracking
+private import semmle.code.java.security.SensitiveActions
+
+/** A sink representing persistent storage that saves data in clear text. */
+abstract class CleartextStorageSink extends DataFlow::Node { }
+
+/** A sanitizer for flows tracking sensitive data being stored in persistent storage. */
+abstract class CleartextStorageSanitizer extends DataFlow::Node { }
+
+/** An additional taint step for sensitive data flowing into cleartext storage. */
+class CleartextStorageAdditionalTaintStep extends Unit {
+  abstract predicate step(DataFlow::Node n1, DataFlow::Node n2);
+}
+
+/** Class for expressions that may represent 'sensitive' information */
+class SensitiveSource extends Expr {
+  SensitiveSource() {
+    // SensitiveExpr is abstract, this lets us inherit from it without
+    // being a technical subclass
+    this instanceof SensitiveExpr
+  }
+
+  /** Holds if this source flows to the `sink`. */
+  cached
+  predicate flowsToCached(Expr sink) {
+    exists(SensitiveSourceFlowConfig conf |
+      conf.hasFlow(DataFlow::exprNode(this), DataFlow::exprNode(sink))
+    )
+  }
+}
+
+/**
+ *  Class representing entities that may be stored/written, with methods
+ *  for finding values that are stored within them, and cases
+ *  of the entity being stored.
+ */
+abstract class Storable extends Call {
+  /** Gets an "input" that is stored in an instance of this class. */
+  abstract Expr getAnInput();
+
+  /** Gets an expression where an instance of this class is stored (e.g. to disk). */
+  abstract Expr getAStore();
+}
+
+private class SensitiveSourceFlowConfig extends TaintTracking::Configuration {
+  SensitiveSourceFlowConfig() { this = "SensitiveSourceFlowConfig" }
+
+  override predicate isSource(DataFlow::Node src) { src.asExpr() instanceof SensitiveExpr }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof CleartextStorageSink }
+
+  override predicate isSanitizer(DataFlow::Node sanitizer) {
+    sanitizer instanceof CleartextStorageSanitizer
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node n1, DataFlow::Node n2) {
+    any(CleartextStorageAdditionalTaintStep c).step(n1, n2)
+  }
+}
+
+private class DefaultCleartextStorageSanitizer extends CleartextStorageSanitizer {
+  DefaultCleartextStorageSanitizer() {
+    this.getType() instanceof NumericType or
+    this.getType() instanceof BooleanType or
+    exists(EncryptedValueFlowConfig conf | conf.hasFlow(_, this))
```

Comment:
This sanitizer seems weird. It removes sinks the have flow from `EncryptedSensitiveMethodAccess`, but that doesn't mean that unencrypted data can't flow to the sink as well. This sort of imprecision might be fine as a heuristic, but why not make the encryption step a sanitizer instead? That seems simpler and more correct. And if we do that, then the question becomes whether or not that makes sense at all, since I don't believe we have default flow through such methods?

Reply:
I had to rethink about these queries again because at first glance it seems that the sanitizer is superfluous, yes.

But the problem is in how the sources (i.e. sensitive expressions) are defined. Since its an heuristic lookup based on the name of the variable, there are cases like the following:

~~~java
String password = request.getParameter("password");
String hPassword = hash(password); // here some hashing function is called, and data-flow doesn't go through
store(hPassword);
~~~

The heuristic `getVariable().getName().toLowerCase().matches("%password%)` picks up both `password` and `hPassword`. And then a data-flow exists between `hPassword` and `store`, making the flow that goes through `hash` irrelevant.

Nevertheless, to fix the (very real) imprecision problem that you highlighted, I changed `EncryptedValueFlowConfig` to have `SensitiveExpr` as its sinks. This means that now the encrypted sources themselves are sanitized, instead of the sinks they reach. I think that this should work better precision-wise. This was implemented in b12a0b7, but let me know if you see a better solution.

Reply:
Yep, that makes more sense.